sunkit_dem/base_model.py

```diff
@@ -0,0 +1,130 @@
+"""
+Base model classes for DEM models
+"""
+from abc import ABC, abstractmethod, abstractclassmethod
+import copy
+
+import numpy as np
+import astropy.units as u
+import astropy.wcs
+import ndcube
+
+
+class BaseModel(ABC):
+
+    @abstractmethod
+    def _model(self):
+        raise NotImplementedError
+
+    @abstractclassmethod
+    def defines_model_for(self):
+        raise NotImplementedError
+
+
+class GenericModel(BaseModel):
+    """
+    Base class for implementing a differential emission measure model
+
+    Parameters
+    ----------
+    data : `~ndcube.NDCubeSequence`
+    kernel : `dict`
+        Each entry should be an `~astropy.units.Quantity` with a key that denotes
+        the wavelength
+    temperature : `~astropy.units.Quantity`
+        Temperature bins over which to compute the DEM
+    """
+
+    _registry = dict()
+
+    def __init_subclass__(cls, **kwargs):
+        """
+        An __init_subclass__ hook initializes all of the subclasses of a given class.
+        So for each subclass, it will call this block of code on import.
+        This replicates some metaclass magic without the need to be aware of metaclasses.
+        Here we use this to register each subclass in a dict that has the
+        `is_datasource_for` attribute.
+        This is then passed into the Map Factory so we can register them.
+        """
+        super().__init_subclass__(**kwargs)
+        if hasattr(cls, 'defines_model_for'):
+            cls._registry[cls] = cls.defines_model_for
+
+    @u.quantity_input
+    def __init__(self, data, kernel, temperature: u.K, **kwargs):
+        self._temperature = self._validate_temperature(temperature)
+        self.data = self._validate_input_data(data)
+        self.kernel = self._validate_kernel(kernel)
+
+    def _validate_temperature(self, temperature):
```

Comment:
An alternative for this would be do use the temperature setter instead of having separate `_validate_*` method.  Some untested code would be:

```Python
    @property
    def temperature(self):
        return self._temperature

    @temperature.setter
    @u.quantity_input(equivalencies=u.temperature_energy())  # allow eV to K conversion because plasma physicists
    def temperature(self, temperature_array: u.K):
        # do validation stuff
        self._temperature = temperature_array
```

Reply:
I like that much better. The `_validate_*` methods were starting to get a bit messy.

Reply:
Looks like I actually fixed this. who knew.